### ISSUE TITLE
Allow multiple DataRightsHolder?

### ISSUE THREAD

**Issue by peterdesmet** [ices-tools-dev/esas#68]
Mentioned by @nicolasvanermen:

> DataRightsHolder we should actually (re)allow multiple values for DataRightsHolder, yet clearly only one value for DataCustodian.

- @cmspinto is this supported by the system?
- It is currently not documented as such in the model: https://esas-docs.ices.dk/tables/#datacustodian

**Comment by cmspinto** [comment]
I agree with Neil, it brings a lot of extra complication to put this field multiple. 
Especially since this list the institution that can release the data, delete or overwrite the data.

**Comment by nicholascarey** [comment]
I think i was assigned this by mistake :)

**Comment by neil-ices-dk** [comment]
> I think i was assigned this by mistake :)

sorry @nicholascarey 😄

**Comment by nicolasvanermen** [comment]
I forwarded this issue to the institutes involved. But regardless of their
answer... when multiple institutes have co-financed certain surveys, they
should both be data rights holder in my opinion. The only complication is
that they need to agree on whether their data should be public or
restricted, but that shouldn't be our problem, or should it?

Nicolas

**Comment by nicolasvanermen** [comment]
To put it differently, assuming you allow only one organisation, they still need to agree on who should be data rights holder, which might be a much more difficult choice...

**Comment by neil-ices-dk** [comment]
> To put it differently, assuming you allow only one organisation, they still need to agree on who should be data rights holder, which might be a much more difficult choice...

i see your point - how do we resolve requests for data in this scenario though:
- All data rights holders would need to be contacted, and all would need to agree to the release. (should this be in parallel, or in serial approval?)

That will need some thinking on how to implement.

**Comment by nicolasvanermen** [comment]
I would say in parallel...

There would still be only one data custodian, and it would be the data
custodian's responsibility to contact the data rights holder(s) and to give
the green light on the release, no? So it seems to me that this process of
(serial or parallel) approval or agreement is somehow masked by the role of
the data custodian...

**Comment by nicolasvanermen** [comment]
To answer the questions assigned to me: we definitely agreed to allow a combination of organizations for the DataRightsHolder field and definitions of DataRightsHolder and DataCustodian are included in our field descriptions in the spreadsheet we are preparing...

**Comment by nicolasvanermen** [comment]
>

**Comment by nicolasvanermen** [comment]
In an ideal world users are associated with only one organisation as a data rights holder (the organisation where they are active at that moment), but with multiple organisations as a data custodian... Yet I think we should leave the possibility to be associated with multiple organisations, what if for example INBO in the future will be fused together with other institutes, and its employees will suddenly be active at an institute with another name?

**Comment by peterdesmet** [comment]
I suggest to not make it needlessly complex and:

1. A user can only be associated with a single organization
2. If a user uploads data, the `DataCustodian` of the submitted data should be the organization associated with the user, otherwise the system returns an error
3. You can upload data with any `DataRightsHolder`(s)
4. You can access data where you are associated with the `DataRightsHolder` (or when the data is public)
5. You can overwrite data where you are associated with the `DataCustodian` or one of the `DataRightsHolder`

---

- So, I (Peter) am associated with INBO (only)
- I can upload data for NIOZ/JNCC (I would only have gotten that data after I received their permission anyway), but I'll have to set the DataCustodian to INBO (makes sense, since that is the org that uploaded it). Let's assume the data is set to private in the system.
- Anyone from NIOZ or JNCC can access the data (makes sense).
- I cannot access the data I uploaded (it's private). This is especially important so that other people from INBO cannot access these data.
- I notice a mistake (on my computer) in the data, so I reupload the data. If the campaignID and both DataRightsHolders match, it is overwritten. Anyone else from INBO could attempt to do the same, but that's very unlikely to work if they don't have the data.
- 3 years later, someone from NIOZ sees an error in one of the campaigns and wants to correct that. Since they are one of the DataRightsHolders, they can upload data for that campaign and overwrite the old info.

**Comment by nicolasvanermen** [comment]
Sounds perfect to me!

**Comment by peterdesmet** [comment]
My only remaining question for overwriting data: should one or all of the DataRightsHolders match to overwrite data?

Let's say the original data is campaignID `10001` with DataRightsHolder `NIOZ~JNCC`. Do I overwrite this campaign if I upload `10001` with `NIOZ~JNCC~INBO` or `NIOZ` (basically updating it with an additional or fewer rightsholder) or are these entirely new campaigns?

**Comment by cmspinto** [comment]
That is fine, and the reason for that is that DataCustodian is not part of the Key.
To find out if a file is a re submission then it is checked if there is another "CampaingID" for the same DatarightsHolder, in a case where the datacustodian changes, it will still detect that is a resubmission.

We will go with the flow you described above.

**Comment by cmspinto** [comment]
> My only remaining question for overwriting data: should one or all of the DataRightsHolders match to overwrite data?
> 
> Let's say the original data is campaignID `10001` with DataRightsHolder `NIOZ~JNCC`. Do I overwrite this campaign if I upload `10001` with `NIOZ~JNCC~INBO` or `NIOZ` (basically updating it with an additional or fewer rightsholder) or are these entirely new campaigns?

No, because like Nicolas and me have hightlighted the key of a campaing are the CampaingID and the datarightsholder and in this case the DataRightsHolder(s) are not the same. So if they find out there was another organization in that campaing they should delete the previous file and submit this new one. 

We can put a warning saying that that campaingID already exists in the datbase with other dataright(s) holder?

**Comment by peterdesmet** [comment]
@cmspinto yeah, such a warning would be nice to have, but not absolutely necessary.

**Comment by cmspinto** [comment]
@peterdesmet: will apply these changes, The format does not changes so it does not affect the data that is already in the database.

**Comment by cmspinto** [comment]
@peterdesmet: will apply these changes, The format does not changes so it does not affect the data that is already in the database.

**Comment by nicolasvanermen** [comment]
We only need to make sure that the sequence in which the multiple data rights holders are given does not matter, so: to identify campaigns 1375~630 should be regarded the same as 630~1375.

**Comment by nicolasvanermen** [comment]
> @peterdesmet: will apply these changes, The format does not changes so it does not affect the data that is already in the database.

Is there already data in the database?

**Comment by cmspinto** [comment]
Yes, this was the first change done, can you test it?
